src/lib/lib.jl

```diff
@@ -207,35 +207,39 @@ end
 
 @generated nt_nothing(x) = Expr(:tuple, [:($f=nothing) for f in fieldnames(x)]...)
 
-@generated pair(::Val{k}, v) where k = :($k = v,)
+@generated pair(::Val{k}, v, _=nothing) where k = :($k = v,)
+@generated pair(::Val{k}, v, ::NamedTuple{keys}) where {k,keys} = k isa Int ? :($(getfield(keys, k)) = v,) : :($k = v,)
 
-@adjoint function literal_getproperty(x, ::Val{f}) where f
-  val = getproperty(x, f)
+@adjoint function literal_getfield(x, ::Val{f}) where f
```

Comment:
Might be cleaner to have separate adjoints for literal_getproperty and literal_getfield

Reply:
Could you explain what you mean here? We now don't define a default adjoint for `literal_getproperty`, but instead fall back to recursing into `getproperty`